File: data_structures/queue/dart/queue.dart
class Queue {
  List<dynamic> queue = [];
  int frontPointer = -1;
  int rearPointer = -1;
  int queueSize;

  Queue(this.queueSize);
  Queue.defalutSize() {
    this.queueSize = 3;
  }

  isFull() {
    if (this.rearPointer == this.queueSize - 1) return true;
    return false;
  }

  isEmpty() {
    if (this.frontPointer == -1 || this.frontPointer > this.rearPointer)
      return true;
    return false;
  }

  enqueue(value) {
    if (isFull()) {
      print("Queue full.");
      return;
    }

    if (this.frontPointer == -1) this.frontPointer++;

    this.queue.add(value);
    ++this.rearPointer;
  }

  int dequeue() {
    if (isEmpty()) {
      print("Queue empty");
      return -1;
    }

    return this.queue[this.frontPointer++];
  }

  printQueue() {
    List<dynamic> result = [];
    for (int i = this.frontPointer; i <= this.rearPointer; i++) {
      result.add(this.queue[i]);
    }
    print(result.join(","));
  }

  front() {
    if (isEmpty()) {
      print("Queue empty");
      return;
    }

    return this.queue[this.frontPointer];
  }

  rear() {
    if (isEmpty()) {
      print("Queue empty");
      return;
    }

    return this.queue[this.rearPointer];
  }
}

void main() {
  Queue queueInstance = Queue.defalutSize();
  queueInstance.enqueue(1);         // [ 1 ]
  queueInstance.enqueue(2);         // [ 1, 2 ]
  queueInstance.enqueue(3);         // [ 1, 2, 3 ]
  queueInstance.enqueue(4);         // Queue full.
  queueInstance.printQueue();       // [ 1, 2, 3 ]
  queueInstance.dequeue();          // 3
  queueInstance.dequeue();          // 2
  queueInstance.dequeue();          // 1
  queueInstance.dequeue();          // Queue empty.
}

File: data_structures/queue/js/queue.js
var Queue = function(queueSize) {
  var queue = [];
  var frontPointer = -1;
  var rearPointer = -1;

  if (queueSize < 0) {
    console.log("Invalid queue size.");
    return;
  }

  var isEmpty = function() {
    if (frontPointer == -1 || frontPointer > rearPointer) return true;
    return false;
  };

  var isFull = function() {
    if (rearPointer == queueSize - 1) return true;
    return false;
  };

  var enqueue = function(value) {
    if (isFull()) {
      console.log("Queue full.");
      return;
    }

    if (frontPointer == -1) {
      frontPointer++;
    }

    queue[++rearPointer] = value;
    console.log(queue);
  };

  var dequeue = function() {
    if (isEmpty()) {
      console.log("Queue empty.");
      return;
    }

    return queue[frontPointer++];
  };

  var front = function() {
    if (isEmpty()) {
      console.log("Queue empty.");
      return;
    }

    return queue[frontPointer];
  };

  var rear = function() {
    if (isEmpty()) {
      console.log("Queue empty.");
      return;
    }

    return queue[rearPointer];
  };

  var printQueue = function() {
    var result = [];
    for (var i = frontPointer; i <= rearPointer; i++) {
      result.push(queue[i]);
    }
    console.log(result);
  };

  return {
    size: queueSize,
    isEmpty: isEmpty,
    isFull: isFull,
    enqueue: enqueue,
    dequeue: dequeue,
    front: front,
    rear: rear,
    printQueue: printQueue
  };
};

var queueInstance = Queue(3);
queueInstance.enqueue(1);           // [ 1 ]
queueInstance.enqueue(2);           // [ 1, 2 ]
queueInstance.enqueue(3);           // [ 1, 2, 3 ]
queueInstance.enqueue(4);           // Queue full.
queueInstance.printQueue();         // [ 1, 2, 3 ]
queueInstance.dequeue();            // 3
queueInstance.dequeue();            // 2
queueInstance.dequeue();            // 1
queueInstance.dequeue();            // Queue empty.
